mypy/dmypy_server.py

```diff
@@ -189,40 +207,31 @@ def serve(self) -> None:
                                 # If we are crashing, report the crash to the client
                                 tb = traceback.format_exception(*sys.exc_info())
                                 resp = {'error': "Daemon crashed!\n" + "".join(tb)}
-                                conn.sendall(json.dumps(resp).encode('utf8'))
+                                server.write(json.dumps(resp).encode('utf8'))
                                 raise
                     try:
-                        conn.sendall(json.dumps(resp).encode('utf8'))
+                        server.write(json.dumps(resp).encode('utf8'))
                     except OSError:
                         pass  # Maybe the client hung up
-                    conn.close()
                     if command == 'stop':
-                        sock.close()
                         reset_global_state()
                         sys.exit(0)
-            finally:
-                # If the final command is something other than a clean
-                # stop, remove the status file. (We can't just
-                # simplify the logic and always remove the file, since
-                # that could cause us to remove a future server's
-                # status file.)
-                if command != 'stop':
-                    os.unlink(STATUS_FILE)
         finally:
-            shutil.rmtree(self.sock_directory)
+            try:
+                server.cleanup()  # try to remove the socket dir on Linux
+            except OSError:
+                pass
+            # If the final command is something other than a clean
+            # stop, remove the status file. (We can't just
+            # simplify the logic and always remove the file, since
+            # that could cause us to remove a future server's
+            # status file.)
+            if command != 'stop':
+                os.unlink(STATUS_FILE)
```

Comment:
I would do this before the cleanup, to avoid leaving a reference to a path that has been deleted?
(Though I doubt it really matters.)